sync/sync.go

```diff
@@ -0,0 +1,345 @@
+		go s.run()
+		return
+	}
+}
+
+func NewSync(peers Peers, layers mesh.Mesh, bv BlockValidator, conf Configuration) *Syncer {
+	s := Syncer{peers, layers, bv, conf, p2p.NewProtocol(peers, protocol, time.Second*5), 0, 0, make(chan bool), make(chan interface{})}
+	s.p.RegisterMsgHandler(LAYER_HASH, s.LayerHashRequestHandler)
+	s.p.RegisterMsgHandler(BLOCK, s.BlockRequestHandler)
+	s.p.RegisterMsgHandler(LAYER_IDS, s.LayerIdsRequestHandler)
+	return &s
+}
+
+//fires a sync every sm.config.syncInterval or on force space from outside
+func (s *Syncer) run() {
+	syncTicker := time.NewTicker(s.config.syncInterval)
+	for {
+		doSync := false
+		select {
+		case <-s.exit:
+			log.Debug("run stoped")
+			return
+		case doSync = <-s.forceSync:
+		case <-syncTicker.C:
+			fmt.Println("timer !!!")
+			doSync = true
+		default:
+			doSync = false
+		}
+		if doSync {
+			go func() {
+				if atomic.CompareAndSwapUint32(&s.SyncLock, IDLE, RUNNING) {
+					fmt.Println("do sync")
+					s.Synchronise()
+					atomic.StoreUint32(&s.SyncLock, IDLE)
+				}
+			}()
+		}
+	}
+}
+
+func (s *Syncer) maxSyncLayer() uint32 {
+	if s.layers.LatestKnownLayer() < s.config.hdist {
+		return 0
+	}
+
+	return s.layers.LatestKnownLayer() - s.config.hdist
+}
+
+func (s *Syncer) Synchronise() {
+	for i := s.layers.LocalLayerCount(); i < s.maxSyncLayer(); {
+		i++
+		blockIds := s.GetLayerBlockIDs(i) //returns a set of all known blocks in the mesh
+		ids := make(chan uint32, len(blockIds))
+		output := make(chan Block)
+
+		//bufferd chan for workers  to read from
+
+		for _, id := range blockIds {
+			ids <- id
+		}
+
+		close(ids)
+
+		// each worker goroutine tries to fetch a block iteratively from each peer
+		for i := 0; i < s.config.concurrency; i++ {
+			go func() {
+				for id := range ids {
+					for _, p := range s.peers.GetPeers() {
+						if bCh, err := s.SendBlockRequest(p, id); err == nil {
+							b := <-bCh
+							if s.sv.ValidateBlock(b) { //some validation testing
+								output <- b
+							}
+						}
+
+					}
+				}
+				close(output)
+			}()
+		}
+
+		blocks := make([]*mesh.Block, 0, len(blockIds))
+		for block := range output {
+			log.Debug("add block to layer", block)
+			blocks = append(blocks, mesh.NewExistingBlock(block.GetId(), block.GetLayer(), nil))
+		}
+		log.Debug("add layer ", i)
+
+		s.layers.AddLayer(mesh.NewExistingLayer(i, blocks))
+	}
+
+	log.Debug("synchronise done, local layer index is ", s.layers.LocalLayerCount())
+}
+
+func (s *Syncer) GetLayerBlockIDs(index uint32) []uint32 {
+
+	m := make(map[string]Peer, 20)
+	peers := s.peers.GetPeers()
+	// request hash from all
+
+	for _, p := range peers {
+
+		hash, err := s.SendLayerHashRequest(p, index)
+		if err != nil {
+			log.Debug("could not get layer ", index, " hash from peer")
+			continue
+		}
+		m[string(hash)] = p
+	}
+
+	var res []uint32
+	for _, v := range m {
+		blocksCh, err := s.SendLayerIDsRequest(v, index)
+		blocks := <-blocksCh
```

Comment:
can this channel block forever

Reply:
no, the protocol_server closes stale channels